operator/controllers/labels.go

```diff
@@ -6,40 +6,72 @@ import (
 		case machinelearningv1.TRANSFORMER:
-			svc.Labels[machinelearningv1.Label_transformer] = "true"
+			svc.Labels[machinelearningv1.Label_component] = machinelearningv1.Label_transformer
 		case machinelearningv1.OUTPUT_TRANSFORMER:
-			svc.Labels[machinelearningv1.Label_output_transformer] = "true"
+			svc.Labels[machinelearningv1.Label_component] = machinelearningv1.Label_output_transformer
 		}
 	}
 	if p.Shadow != true && (p.Traffic >= 50 || p.Traffic == 0) {
-		svc.Labels[machinelearningv1.Label_default] = "true"
+		svc.Labels[machinelearningv1.Label_endpoint] = machinelearningv1.Label_default
 	}
 	if p.Shadow == true {
-		svc.Labels[machinelearningv1.Label_shadow] = "true"
+		svc.Labels[machinelearningv1.Label_endpoint] = machinelearningv1.Label_shadow
 	}
 	if p.Traffic < 50 && p.Traffic > 0 {
-		svc.Labels[machinelearningv1.Label_canary] = "true"
+		svc.Labels[machinelearningv1.Label_endpoint] = machinelearningv1.Label_canary
 	}
 	if !isEmptyExplainer(p.Explainer) {
-		svc.Labels[machinelearningv1.Label_explainer] = "true"
+		svc.Labels[machinelearningv1.Label_endpoint] = machinelearningv1.Label_explainer
 	}
 	svc.Labels[machinelearningv1.Label_managed_by] = machinelearningv1.Label_value_seldon
 }
 
-func addLabelsToDeployment(deploy *appsv1.Deployment, containerServiceKey, containerServiceValue string) {
-	deploy.ObjectMeta.Labels[containerServiceKey] = containerServiceValue
-	deploy.Spec.Selector.MatchLabels[containerServiceKey] = containerServiceValue
-	deploy.Spec.Template.ObjectMeta.Labels[containerServiceKey] = containerServiceValue
+func addLabelsToDeployment(deploy *appsv1.Deployment, pu *machinelearningv1.PredictiveUnit, p *machinelearningv1.PredictorSpec) {
+	if pu.Type != nil {
+		switch *pu.Type {
+		case machinelearningv1.ROUTER:
+			deploy.Labels[machinelearningv1.Label_component] = machinelearningv1.Label_router
+			deploy.Spec.Template.ObjectMeta.Labels[machinelearningv1.Label_component] = machinelearningv1.Label_router
+		case machinelearningv1.COMBINER:
+			deploy.Labels[machinelearningv1.Label_component] = machinelearningv1.Label_combiner
+			deploy.Spec.Template.ObjectMeta.Labels[machinelearningv1.Label_component] = machinelearningv1.Label_combiner
+		case machinelearningv1.MODEL:
+			deploy.Labels[machinelearningv1.Label_component] = machinelearningv1.Label_model
+			deploy.Spec.Template.ObjectMeta.Labels[machinelearningv1.Label_component] = machinelearningv1.Label_model
+		case machinelearningv1.TRANSFORMER:
+			deploy.Labels[machinelearningv1.Label_component] = machinelearningv1.Label_transformer
+			deploy.Spec.Template.ObjectMeta.Labels[machinelearningv1.Label_component] = machinelearningv1.Label_transformer
+		case machinelearningv1.OUTPUT_TRANSFORMER:
+			deploy.Labels[machinelearningv1.Label_component] = machinelearningv1.Label_output_transformer
+			deploy.Spec.Template.ObjectMeta.Labels[machinelearningv1.Label_component] = machinelearningv1.Label_output_transformer
+		}
+	}
```

Comment:
You can usually define multiple components of the graph as containers running inside the same pod / deployment`. In that case, wouldn't `addLabelsToDeployment` just override the previous label for each component?

Reply:
Thanks @adriangonz this is a good point. I'll look at adding the old labels back in. I think there are still some useful changes in this PR which add labels where before none were being added. e.g.

https://github.com/SeldonIO/seldon-core/pull/2130/files/cf55b7102a0a44b358f82457452d0792a0e5be94#diff-b520473967af374ae98ad6019e5b84a4R663

and

https://github.com/SeldonIO/seldon-core/pull/2130/files/cf55b7102a0a44b358f82457452d0792a0e5be94#diff-b520473967af374ae98ad6019e5b84a4R901

So I'll go through and cherry-pick the useful changes and discard the redundant ones.

Reply:
Fixed. Labels should aggregate now instead of overriding each other.